src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+
+    @property
+    def environment(self):
+        """
+        Get (entire) environment of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["environment"]
+        except KeyError:
+            return
+
+    @environment.setter
+    def environment(self, environ):
+        """
+        Set (entire) environment of job.
+        """
+        if not isinstance(environ, collectionsAbc.Mapping):
+            raise ValueError("environment must be a mapping")
+        self.jobspec["attributes"]["system"]["environment"] = environ
+
+    def _set_treedict(self, in_dict, key, val):
+        """
+        _set_treedict(d, "a.b.c", 42) is like d[a][b][c] = 42
+        but levels are created on demand.
+        """
+        path = key.split(".", 1)
+        if len(path) == 2:
+            self._set_treedict(in_dict.setdefault(path[0], {}), path[1], val)
+        else:
+            in_dict[key] = val
+
+    def setattr(self, key, val):
+        """
+        set job attribute
+        """
+        self._set_treedict(self.jobspec, "attributes." + key, val)
+
+    def setattr_shopt(self, key, val):
+        """
+        set job attribute: shell option
+        """
+        self.setattr("system.shell.options." + key, val)
+
+    def dumps(self):
+        return json.dumps(self.jobspec)
+
+    @property
+    def resources(self):
+        return self.jobspec.get("resources", None)
+
+    @property
+    def tasks(self):
+        return self.jobspec.get("tasks", None)
+
+    @property
+    def attributes(self):
+        return self.jobspec.get("attributes", None)
+
+    @property
+    def version(self):
+        return self.jobspec.get("version", None)
+
+    def resource_walk(self):
+        """
+        Traverse the resources in the `resources` section of the jobspec.
+
+        Performs a depth-first, pre-order traversal.
+        """
+
+        def walk_helper(res_list):
+            for resource in res_list:
+                yield resource
+                children = resource.get("with", [])
+                # TODO: convert to `yield from` after dropping 2.7
+                for res in walk_helper(children):
+                    yield res
+
+        return walk_helper(self.resources)
+
+    def resource_walk_type(self, res_type):
+        """
+        Traverse the resources of type `res_type` in the `resources` section of the
+        jobspec.
+
+        Performs a depth-first, pre-order traversal.
+
+        :param res_type: resource type(s) to filter by
+        :type res_type: string or iterable of strings
+        """
+        if isinstance(res_type, six.string_types):
+            func = lambda x: x["type"] == res_type
+        elif isinstance(res_type, collectionsAbc.Iterable):
+            func = lambda x: x["type"] in res_type
+        else:
+            raise TypeError("res_type must be a string or an iterable")
+        return (x for x in self.resource_walk() if func(x))
+
+    def count_num_type(self, res_type):
```

Comment:
This makes me think we could really do with an additional walk-like generator method to give out aggregates of all resources of a type. It would be nice if this could be done iteratively, but without some kind of helper to create a generator for it there's no good way to do it, will have to give this some thought.

Reply:
Sounds good. 

> to give out aggregates of all resources of a type

By aggregate, are you thinking of a dict whose keys are resource types and values are counts of those resources, or where values are a list of all resources of that type?

Reply:
Dict-like is what I was thinking, but either might be useful.

Reply:
I was going down the path of creating a method to produce the counts of resources, and I ran face first into the `unit` key.  It isn't clear from RFC14 which units are valid, how to convert between them, etc.  So for now, I punted on that problem and just ignore the unit and sum the counts as-is.  If we keep the new counting method in this PR, we can open an issue to track this.  An initial first improvement could be to convert between B, KB, MB, GB, etc.

Reply:
Oof, yeah, that's a good point.  We'll have to get more specific about units and conversions for some things like that.